crates/sui-core/src/authority_server.rs

```diff
@@ -371,17 +371,18 @@ impl ValidatorService {
             return Ok(tonic::Response::new(response.into()));
         }
 
-        // 2) Verify cert signatures
+        // 2) If the validator is already halted due to reconfiguration, we stop here, to avoid
+        // sending the transaction to consensus.
+        fp_ensure!(
+            state.should_accept_user_certs().await,
```

Comment:
this is not a new issue in your PR, but this has me thinking: Have we had a race here this whole time? Consider:

- handle_cert request passes this check (or passes the validator halted check in the old code)
- in another thread, we halt the validator for reconfig
- for some reason the handle_cert request thread pauses for a long time
- we finish executing all certs in final checkpoint
- we check if there are extra txes to revert, there are none.
- the handle_cert request thread wakes up and finishes executing.
- now we are forked

admittedly this is far fetched, but its not impossible - especially if the execution thread has to fetch objects from cold storage or distributed storage in the future.

I think to really be correct, we can't halt the validator until all current executions have finished. the simple way to enforce this would be:
- During tx execution: acquire a read lock on the reconfig state, check if halted, hold read lock until execution is finished.
- During reconfig, acquire a write lock before updating the reconfig state.

another detail we'll have to consider: There could be partially written txns (managed by the current WAL or by sally) which won't show up as extra transactions if they haven't finished committing. this state will have to be cleaned up.

Reply:
meant to add also that the simple way i mentioned may not be suitable for high parallelism, i don't know how many-reader contention on an rw lock behaves in practice. But we could build this functionality into mutex table. Reconfig would have to acquire a lock on every entry in the mutex table that is exclusive with any existing tx locks.

Reply:
Yeah all great points! I started to think about them too.
The existing code doesn't have data race because we also gate it when getting a ticket from the notifier.
The new code certainly does.
I need to talk to Andey regarding where we would be making the call. Perhaps it needs to be in a central place so that everything is synchrounized.

Reply:
> The existing code doesn't have data race because we also gate it when getting a ticket from the notifier.

but what if we halt the validator right after the ticket is issued, but before execution is finished? it seems like the same race can happen.

anyway, lets file a bug for the race, i'm ok with not fixing this right now unless you can think of a way to do it quickly.

Reply:
I don't think we need to file a bug. In the existing code, we also wait for the ticket system to drain before closing the door, just for the scenario you described.

Now in the new code, I think it would be easier to deal with. Below is the plan:
I will turn this into a RwLock.
There will be a central place in consensus adapter to send transactions to consensus.
Before sending the last message ton consensus, we will acquire a write lock, and only release it after sending it; before sending a normal transaction, we will acquire a read lock, and only release it after sending it. This should be bullet proof.

Reply:
I think I am ok with TODOs there - this is something we are going to address in a near term